src/rust/grapl-utils/src/rusoto_ext/dynamodb.rs

```diff
@@ -125,55 +121,175 @@ pub trait GraplDynamoDbClientExt: DynamoDb + Send + Sync {
-                }
-                _ => {}
+                pending_items.extend(unprocessed_keys);
             }
         }
 
-        Ok(BatchGetItemOutput {
+        let output = BatchGetItemOutput {
             consumed_capacity: batch_get_items_input
                 .return_consumed_capacity
                 .map(|_| total_consumed_capacity),
             responses: Some(total_responses),
             unprocessed_keys: None,
-        })
+        };
+
+        Ok(output)
+    }
+
+    /// Similar to [`batch_get_item_reliably`], this function handles chunking requests to avoid going
+    /// over limits imposed on the size of BatchWriteItem operations in DynamoDB.
+    async fn batch_write_item_reliably(
+        &self,
+        batch_write_item: BatchWriteItemInput
+    ) -> Result<BatchWriteItemOutput, RusotoError<BatchWriteItemError>> {
+        let BatchWriteItemInput {
+            request_items,
+            return_consumed_capacity,
+            return_item_collection_metrics
+        } = batch_write_item;
+
+        let mut pending_writes: Vec<_> = request_items.into_iter()
+            .flat_map(|(table_name, write_requests)| {
+                write_requests.into_iter()
+                    .map(|write_request| (table_name.clone(), write_request))
+                    .collect::<Vec<_>>()
+            }).collect();
+
+        let mut total_consumed_capacity = Vec::new();
+        let mut item_collection_metrics = HashMap::new();
+
+        while !pending_writes.is_empty() {
+            let mut writes_to_process = HashMap::new();
+
+            pending_writes
+                .drain(0 ..std::cmp::min(pending_writes.len(), DYNAMODB_MAX_BATCH_WRITE_ITEM_SIZE))
+                .for_each(|(table_name, write_request)| {
+                    writes_to_process
+                        .entry(table_name)
+                        .or_insert(Vec::new())
+                        .push(write_request);
+                });
+
+            let batch_write_operation = BatchWriteItemInput {
+                request_items: writes_to_process,
+                return_consumed_capacity: return_consumed_capacity.clone(),
+                return_item_collection_metrics: return_item_collection_metrics.clone()
+            };
+
+            let batch_write_response = self.batch_write_item(batch_write_operation).await?;
+
+            // record consumed capacity, if enabled
+            if let Some(capacity) = batch_write_response.consumed_capacity {
+                total_consumed_capacity.extend(capacity);
+            }
+
+            if let Some(metrics) = batch_write_response.item_collection_metrics {
+                metrics.into_iter()
+                    .for_each(|(table_name, item_metrics)| {
+                        item_collection_metrics.entry(table_name)
+                            .or_insert(Vec::new())
+                            .extend(item_metrics);
+                    });
+            }
```

Comment:
Feels like maybe what we want is some sort of BatchWriteMetricsWriter or something, which would just take a response and handle all of this stuff. That way our metrics / capacity code isn't clogging up our business logic.

Reply:
I don't understand this point. This isn't business logic; it's specifically for scaling up and handling dynamodb's request limits transparently.